common/src/main/java/io/pravega/common/concurrent/FutureHelpers.java

```diff
@@ -263,6 +263,33 @@
     }
 
     /**
+     * Same as CompletableFuture.exceptionally(), except that it allows returning a CompletableFuture instead of a single value.
+     *
+     * @param future  The original CompletableFuture to attach an Exception Listener.
+     * @param handler A Function that consumes a Throwable and returns a CompletableFuture of the same type as the original one.
+     *                This Function will be invoked if the original Future completed exceptionally.
+     * @param <T>     Type of the value of the original Future.
+     * @return A new CompletableFuture that will be completed either with the result of future (if it completed normally),
+     * or with the result of handler when applied to the exception of future, should future complete exceptionally.
+     */
+    public static <T> CompletableFuture<T> exceptionallyCompose(CompletableFuture<T> future, Function<Throwable, CompletableFuture<? extends T>> handler) {
+        CompletableFuture<T> result = new CompletableFuture<>();
```

Comment:
I think a much simpler implementation of this function would be:
```
           future.handle((u, t) -> {
               if (t == null) {
                   return CompletableFuture.completedFuture(u);
               }
               return handler.apply(t);
           }).thenCompose(f -> f);
```

Reply:
`handle` does not allow returning a `CompletableFuture`. It returns the same type as the one of the original `future`. It has the same signature as `exceptionally`, hence the need for this complication.

Reply:
This code works I tested it. Handle can return any type.